src/java/map/BinaryMask.java

```diff
@@ -93,6 +96,41 @@ public void set(int x, int y, boolean value) {
         mask[x][y] = value;
     }
 
+    public float getDistanceBetweenPoints(int x1, int y1, int x2, int y2) {
+        return (float) StrictMath.sqrt((double) ((x1 - x2) ^ 2 + (y1 - y2) ^ 2));
+    }
+
+    public float getDistanceBetweenPoints(int x1, int y1, Point Point2) {
+        return (float) StrictMath.sqrt((double) ((x1 - Point2.x) ^ 2 + (y1 - Point2.y) ^ 2));
+    }
+
+    public float getDistanceBetweenPoints(Point Point1, Point Point2) {
+        return (float) StrictMath.sqrt((double) ((Point1.x - Point2.x) ^ 2 + (Point2.x - Point2.y) ^ 2));
+    }
+
+    public Point getNearestPointInAreaTo(int x, int y) {
+        Point nearestPoint = new Point (x, y);
+        int rank;
+        FloatMask distanceField = getDistanceFieldFromPointInArea(x, y);
+        for (int a = 0; a < getSize(); a++) {
```

Comment:
You should check if the point is n area then just return point if it is. If it is not in area you only need to check the edges of area not the entire mask